File: exam2/src/java/SameColorTest.java
import java.util.Scanner;
import java.util.ArrayList;
import java.util.List;
import java.io.*;

class SameColorException extends Exception {
    public SameColorException(String errMsg) {
        super(errMsg);
    }
}


public class SameColorTest extends Test {
    /**
     * runTest: run the test in the specified file.  Returns true, if
     *   the test passes and false otherwise.
     **/
    public boolean runTest(File f, boolean verbose) {
        Scanner scanner = null;
        try {
            scanner = new Scanner(f);
        } catch (FileNotFoundException e) {
            if (verbose) {
                System.out.println("Results file not found:" + f.getPath());
            }
            return false;
        }

        Graph g = loadGraph(scanner, f.getPath(), verbose);

        String source_name = null;
        String dest_name = null;        
        try {
            String line = scanner.nextLine();
            String[] data = line.split(" ");
            source_name = data[0];
            dest_name = data[1];            
        } catch (Exception e) {
            if (verbose) {
                String msg = "Unexpected exception when parsing the last line of the input file: ";
                System.out.println(msg + e);
            }
            return false;
        }

        Boolean expected = getResults(f, verbose);
        if (expected == null) {
            // bad results file
            return false;
        }

        Boolean actual = null;
        try {
            actual = SameColor.hasSameColor(g, source_name, dest_name);
        } catch (Exception e) {
            if (verbose) {
                System.out.println("Unexpected exception from call to hasSameColor: " + e);        
            }
            return false;
        }

        return expected == actual;
    }


    /**
     * loadGraph: load a graph from a file
     **/
    private Graph loadGraph(Scanner scanner, String filename, boolean verbose) {
        Graph g = new Graph();
        String line = "";
        try {
            line = scanner.nextLine();
            String[] data = line.split(" ");
            int numVertices = Integer.parseInt(data[0]);
            int numEdges = Integer.parseInt(data[1]);            

            for (int i = 0; i < numVertices; i++) {
                // vertex format: name color
                line = scanner.nextLine();
                data = line.split(" ");
                Vertex v = new Vertex(data[0], data[1]);
                g.addVertex(v);
            }

            for (int i = 0; i < numEdges; i++) {
                // edge format: source-vertex-name dest-vertex-name
                line = scanner.nextLine();
                data = line.split(" ");
                Vertex source = g.getVertex(data[0]);
                Vertex dest = g.getVertex(data[1]);
                if (source == null) {
                    throw new SameColorException("Bad source vertex: " + data[0]);

                }
                if (dest == null) {
                    throw new SameColorException("Bad destination vertex: " + data[1]);
                }
                source.addEdge(dest);
            }
        } catch (NumberFormatException e) {
            if (verbose) {
                System.out.println("Format error: line:" + line);
            }
            return null;
        } catch (Exception e) {
            if (verbose) {
                System.out.println("Unexpected exception: " + e);
            }
            return null;
        }

        return g;
    }

    /**
     * getResults: get the expected results from the results file.
     **/
    private static Boolean getResults(File inputFile, boolean verbose) {
        Boolean result = null;
        Scanner scanner = TestHelpers.getResultsFileScanner(inputFile);

        if (scanner == null) {
            return result;
        }

        try {
            String data = scanner.nextLine();
            result = data.toLowerCase().equals("true");
        } catch (Exception e) {
            if (verbose) {
                System.out.println("Expected exception when parsing the expected results file: " + e);
            }
            return null;
        }
        return result;
    }
    
    public static void main(String[] args) {
        System.exit(TestHelpers.doTests(args, new SameColorTest()));
    }
}

    

File: exam2/src/cpp/SameColorTest.cpp
#include "TestHelpers.h"
#include "SameColor.h"
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <assert.h>

namespace test {

bool get_result(std::string& f) {
    std::string resultsFileName = TestHelpers::getResultsFile(f);
    std::ifstream ifs(resultsFileName);
    std::string res = "";

    if (ifs) {
        ifs >> res;
        if (res == "") {
            std::cout << "Format error" << std::endl;
        }
    } else {
        std::cout << "Results file not found: " << resultsFileName << std::endl;
        exit(1);
    }
    return res == "True";
 }

void get_data(
    std::string& f,
    Graph* g,
    std::string& source_name,
    std::string& dest_name
    ) {
    std::string dataFileName = TestHelpers::getDataFile(f);
    int numVertices = -1, numEdges = -1;

    std::ifstream ifs(dataFileName);
    if (ifs) {
        ifs >> numVertices >> numEdges;
        if (numVertices == -1 || numEdges == -1) {
            std::cout << "Format error" << std::endl;
            return;
        }

        for (size_t i = 0; i < numVertices; i++) {
            std::string name, color;
            ifs >> name >> color;
            g->add_vertex(new Vertex(name, color));
        }

        for (size_t i = 0; i < numEdges; i++) {
            // edge format: source-vertex-name dest-vertex-name
            std::string source_name, dest_name;
            ifs >> source_name >> dest_name;
            Vertex* source = g->get_vertex(source_name);
            Vertex* dest = g->get_vertex(dest_name);
            if (source == nullptr) std::cout << "Bad source vertex  " << source_name << "\n";
            if (dest == nullptr) std::cout << "Bad destination vertex  " << dest_name << "\n";
            source->addEdge(dest);
        }
        ifs >> source_name >> dest_name;
    } else {
        std::cout << dataFileName << " failed to be read" << std::endl;;
        return;
    }

    return;
}


bool runTest(std::string f) {
    Graph* g = new Graph();
    std::string source_name, dest_name;
    get_data(f, g, source_name, dest_name);

    int actual = has_same_color_path(g, source_name, dest_name);
    int expected = get_result(f);

    if (actual == expected) {
        return true;
    } else {
        std::cout << "Failed actual result and expected result do not match" << "\n";
        std::cout << "    actual is " << std::to_string(actual) << "\n";
        std::cout << "    expected is " << std::to_string(expected) << "\n";
	    return false;
    }

    return true;
}

} // namespace test

int main (int argc, char* argv[])
{
    std::vector<std::string> files = TestHelpers::getFiles(argc, argv);
    if (files.size() == 0) {
        std::cout << "no test file found";
        exit(1);
    }

    int count = 0;
    for (size_t i = 0; i < files.size(); ++i) {
        std::string f = files[i];
        std::cout << "Test " << f << ": ";
        if (test::runTest(f)) {
            std::cout << "PASSED\n";
            count++;
        } else {
            std::cout << "FAILED\n";
        }
    }

    if (count == files.size()) {
        return 42;
    } else {
        return 0;
    }

}
